File: src/mods/town-features.ts
// WHAT A BUILDING DOES — the town-feature table, and the rows a faction adds.
//
// Every compiled effect a special building has is a FEATURE, a number the
// executable keys on: the Library's extra spell, the Hall of Trial's three
// warcry tiers, the Capitol. One table at 0x10909B0 (`FEATURE_TABLE_RVA`)
// says which building of which town grants which, at what level —
// forty-seven rows, `{feature, town, building, minLevel}`, town 2 for any.
// The town's readers scan it (`native/faction/town-features.c` lists them),
// `OnBuildingBuilt` asks it what was just built, and a building the table
// does not know does nothing whatever: a hall of ours with three levels
// stood empty in launch 35 for that reason alone.
//
// The extension owns the table now — the engine's rows and then ours, read
// from `feature` lines of `bin/homm5-editor-buildings.txt`:
//
//   feature <townType> <buildingType> <minLevel> <feature>
//
// and a building of ours grants a feature by naming the SHIPPED building
// whose effect it wants (`BuildingEdit.grants: { like: 'TOWN_STRONGHOLD' }`):
// the rows below are copied level for level under our town and our slot.
// A feature is what the executable does with the number — the forty-six
// effects the eight towns ship — so this is the whole menu; an effect of
// our own is a term of the extension's, not a row here.
//
// `SHIPPED_FEATURES` is the table as read on 2026-09-18; `test-town-features`
// reads it back out of the executable.

import { buildingOrdinal } from './town-button.ts';

/** Where the compiled table sits in H5_Game_H5E.exe, as an RVA. */
export const FEATURE_TABLE_RVA = 0xc909b0;

export interface FeatureRow {
  /** The feature's number — what the executable's effect code keys on. */
  feature: number;
  /** A `TownType` ordinal; 2 (TOWN_NO_TYPE) for any town. */
  town: number;
  /** A `TownBuildingType` ordinal. */
  building: number;
  /** The building's level that grants it, 1-based. */
  minLevel: number;
}

const T = { ANY: 2, HAVEN: 3, SYLVAN: 4, ACADEMY: 5, DUNGEON: 6, NECRO: 7, INFERNO: 8, FORTRESS: 9, STRONGHOLD: 10 };
const S = (n: number): number => 16 + n;

/** The forty-seven rows of the compiled table, in its order. */
export const SHIPPED_FEATURES: readonly FeatureRow[] = [
  { feature: 0x00, town: T.HAVEN, building: S(1), minLevel: 1 },
  { feature: 0x01, town: T.HAVEN, building: S(2), minLevel: 1 },
  { feature: 0x02, town: T.HAVEN, building: S(3), minLevel: 1 },
  { feature: 0x03, town: T.HAVEN, building: S(4), minLevel: 1 },
  { feature: 0x04, town: T.HAVEN, building: S(5), minLevel: 1 },
  { feature: 0x05, town: T.INFERNO, building: S(1), minLevel: 1 },
  { feature: 0x06, town: T.INFERNO, building: S(2), minLevel: 1 },
  { feature: 0x07, town: T.INFERNO, building: S(3), minLevel: 1 },
  { feature: 0x08, town: T.INFERNO, building: S(4), minLevel: 1 },
  { feature: 0x09, town: T.INFERNO, building: S(5), minLevel: 1 },
  { feature: 0x0a, town: T.SYLVAN, building: S(0), minLevel: 1 },
  { feature: 0x0b, town: T.SYLVAN, building: S(0), minLevel: 2 },
  { feature: 0x0c, town: T.SYLVAN, building: S(2), minLevel: 1 },
  { feature: 0x0d, town: T.SYLVAN, building: S(2), minLevel: 2 },
  { feature: 0x0e, town: T.SYLVAN, building: S(4), minLevel: 1 },
  { feature: 0x0f, town: T.SYLVAN, building: S(5), minLevel: 1 },
  { feature: 0x10, town: T.NECRO, building: S(1), minLevel: 1 },
  { feature: 0x11, town: T.NECRO, building: S(3), minLevel: 1 },
  { feature: 0x12, town: T.NECRO, building: S(2), minLevel: 1 },
  { feature: 0x13, town: T.NECRO, building: S(4), minLevel: 1 },
  { feature: 0x14, town: T.NECRO, building: S(5), minLevel: 1 },
  { feature: 0x15, town: T.ACADEMY, building: S(1), minLevel: 1 },
  { feature: 0x16, town: T.ACADEMY, building: S(2), minLevel: 1 },
  { feature: 0x17, town: T.ACADEMY, building: S(3), minLevel: 1 },
  { feature: 0x18, town: T.ACADEMY, building: S(4), minLevel: 1 },
  { feature: 0x19, town: T.ACADEMY, building: S(5), minLevel: 1 },
  { feature: 0x1a, town: T.DUNGEON, building: S(1), minLevel: 1 },
  { feature: 0x1b, town: T.DUNGEON, building: S(1), minLevel: 2 },
  { feature: 0x1c, town: T.DUNGEON, building: S(3), minLevel: 1 },
  { feature: 0x1d, town: T.DUNGEON, building: S(4), minLevel: 1 },
  { feature: 0x1e, town: T.DUNGEON, building: S(5), minLevel: 1 },
  { feature: 0x1f, town: T.DUNGEON, building: S(6), minLevel: 1 },
  { feature: 0x20, town: T.FORTRESS, building: S(1), minLevel: 1 },
  { feature: 0x21, town: T.FORTRESS, building: S(1), minLevel: 2 },
  { feature: 0x22, town: T.FORTRESS, building: S(1), minLevel: 3 },
  { feature: 0x23, town: T.FORTRESS, building: S(2), minLevel: 1 },
  { feature: 0x24, town: T.FORTRESS, building: S(3), minLevel: 1 },
  { feature: 0x25, town: T.FORTRESS, building: S(4), minLevel: 1 },
  { feature: 0x26, town: T.FORTRESS, building: S(5), minLevel: 1 },
  { feature: 0x27, town: T.STRONGHOLD, building: S(1), minLevel: 1 },
  { feature: 0x28, town: T.STRONGHOLD, building: S(1), minLevel: 2 },
  { feature: 0x29, town: T.STRONGHOLD, building: S(1), minLevel: 3 },
  { feature: 0x2a, town: T.STRONGHOLD, building: S(2), minLevel: 1 },
  { feature: 0x2b, town: T.STRONGHOLD, building: S(3), minLevel: 1 },
  { feature: 0x2c, town: T.STRONGHOLD, building: S(4), minLevel: 1 },
  { feature: 0x2d, town: T.STRONGHOLD, building: S(5), minLevel: 1 },
  { feature: 0x2e, town: T.ANY, building: 0, minLevel: 4 },
];

/** The shipped town ordinals the table keys on, by `TownType` name. */
export const FEATURE_TOWNS: Readonly<Record<string, number>> = {
  TOWN_HEAVEN: T.HAVEN, TOWN_PRESERVE: T.SYLVAN, TOWN_ACADEMY: T.ACADEMY, TOWN_DUNGEON: T.DUNGEON,
  TOWN_NECROMANCY: T.NECRO, TOWN_INFERNO: T.INFERNO, TOWN_FORTRESS: T.FORTRESS, TOWN_STRONGHOLD: T.STRONGHOLD,
};

/** What a building of ours grants: the effect of a shipped town's building, level for level. */
export interface Grant {
  /** The shipped town — `TOWN_STRONGHOLD`. */
  like: string;
  /** Its building; the same type as ours when absent. */
  building?: string;
}

/** A row of ours before the town is numbered: the copier knows the type's name, not its ordinal. */
export interface OwnFeature {
  building: number;
  minLevel: number;
  feature: number;
}

/**
 * The rows a building of ours gets from a grant — every shipped row of that
 * town and building, at its own level. None is a refusal — the building
 * named has no compiled effect — unless the grant is only implied by `from`.
 */
export function grantedFeatures(ourBuilding: string, grant: Grant, what: string, implied = false): OwnFeature[] {
  const town = FEATURE_TOWNS[grant.like];
  if (!town) throw new Error(`${what}: grants like ${grant.like}, which is not a shipped town`);
  const source = buildingOrdinal(grant.building ?? ourBuilding);
  const rows = SHIPPED_FEATURES.filter((r) => r.town === town && r.building === source);
  // A building taken whole brings what it had — a guild's stubs, nothing; asked for outright, nothing is a mistake.
  if (!rows.length && !implied) throw new Error(`${what}: ${grant.like}'s ${grant.building ?? ourBuilding} has no effect to grant`);
  const ours = buildingOrdinal(ourBuilding);
  return rows.map((r) => ({ building: ours, minLevel: r.minLevel, feature: r.feature }));
}

/** The rows' lines for the buildings file, the town numbered. */
export function featureLines(town: number, rows: readonly OwnFeature[]): string[] {
  return rows.map((r) => `feature ${town} ${r.building} ${r.minLevel} ${r.feature}`);
}

File: native/faction/town-features.c
// What a building DOES — the town-feature table, ours, with the engine's rows in it.
//
// A piece of the ONE translation unit — native/homm5-editor.c includes every
// piece in order, so everything included before this file is visible here and
// nothing after it is. Statics stay statics; nothing here is a module.

/** Which switch turns this file's logging on — see the bottom of core/log.c. */
#undef LOG_UNIT
#define LOG_UNIT faction_town_features

// ---------------------------------------------------------------------------
// WHAT THE ENGINE HAS. Every compiled effect a special building has — the
// Library's extra spell, the Hall of Trial's three warcry tiers, the Capitol
// — is a FEATURE, a number 0…0x2E, and one table at 0x10909B0 says which
// building of which town grants it:
//
//   struct { int feature; int town; int building; int minLevel; }  ×47
//
// `town` 2 (TOWN_NO_TYPE) means any: the Capitol row (0x2E, TB_TOWN_HALL ≥ 4)
// is the one such. The rest are the eight towns' specials, one row per
// (town, slot, level) — Sylvan's SPECIAL_0 at levels 1 and 2 are two features,
// Stronghold's SPECIAL_1 at levels 1, 2, 3 are 0x27, 0x28, 0x29. Six small
// readers of CAdvMapTown scan it, all with the same test — row.town == 2 or
// row.town == GetType() (slot +0xD4) — and they are vtable slots +0x40 … +0x54:
//
//   +0x40  0xAC8970  may this feature be built here (row → +0x5C, +0xB4)
//   +0x44  0xAC8A10  HasFeature(f):  BuildingLevel(row.building) >= row.minLevel
//   +0x48  0xAC8A80  FeatureOf(building): the FIRST row's feature, else -1
//   +0x4C  0xAC8AD0  BuildingOf(f): the row's building, else 0x1A
//   +0x50  0xAC8B20  a feature's building through RowOf (0xAD0710), by index
//   +0x54  0xAC8B90  as +0x40, another verdict
//
// and `OnBuildingBuilt` (0xAC7FF0) asks FeatureOf and, for 0x27 at level 0,
// fills the town's warcry list (0xAC4740) — which is why a hall of ours stood
// empty in launch 35: three levels built, no row, no feature, no warcries.
//
// WHAT WE DO. The table is OURS: the engine's 47 rows copied, then a row per
// `feature` line of `bin/homm5-editor-buildings.txt`:
//
//   feature <townType> <buildingType> <minLevel> <feature>
//
// and every one of the twenty places the six readers name the engine's table
// — its start, its end, a column — is pointed at ours instead, at the same
// offset from the start. The readers' loops, tests and index arithmetic stay
// the engine's; only the address moves. Nothing in the executable is
// extended in place: the compiled table stays as it is, unread.
//
// The order matters and is kept: FeatureOf answers the FIRST row for a
// (town, building), so a hall's rows go in as 0x27, 0x28, 0x29 like
// Stronghold's, and the editor writes them so. The engine's rows come first,
// so a town of ours never shadows a shipped one — its type matches no shipped
// row but the Capitol's, which is any town's.
//
// The address bytes of those twenty references are the loader's to rewrite
// under ASLR, so each is checked against the table's address as loaded — the
// opcode bytes before it against what was read — and refused, all twenty or
// none, when a build differs.

#define FEATURE_TABLE_RVA 0xc909b0u
#define SHIPPED_FEATURE_ROWS 47
#define MAX_OWN_FEATURE_ROWS 64

typedef struct {
  int feature;
  int town;
  int building;
  int minLevel;
} FeatureRow;

static FeatureRow g_ownFeatures[MAX_OWN_FEATURE_ROWS];
static int g_ownFeatureCount = 0;

static void load_town_features(void) {
  DWORD got = 0;
  char *buf = read_beside_us(L"homm5-editor-buildings.txt", &got);
  if (!buf) return;
  const char *p = buf, *end = buf + got;
  while (p < end) {
    const char *line = p;
    while (p < end && *p != '\n') p++;
    const char *stop = p;
    if (p < end) p++;
    while (line < stop && (*line == ' ' || *line == '\t')) line++;
    if (line >= stop || *line == '#') continue;
    const char *q = line;
    if (!take_word(&q, stop, "feature")) continue;
    FeatureRow r;
    if (!read_int(&q, stop, &r.town) || !read_int(&q, stop, &r.building)) continue;
    if (!read_int(&q, stop, &r.minLevel) || !read_int(&q, stop, &r.feature)) continue;
    if (r.town < 3 || r.building < 0 || r.minLevel < 1 || r.feature < 0) continue;
    if (g_ownFeatureCount < MAX_OWN_FEATURE_ROWS) g_ownFeatures[g_ownFeatureCount++] = r;
  }
  VirtualFree(buf, 0, MEM_RELEASE);
  log_num("town features: rows of ours: ", g_ownFeatureCount);
}

// ---------------------------------------------------------------------------
// The twenty references: where each instruction sits, the opcode bytes before
// its address operand, and what the operand points at as an offset into the
// table (0x2F0 is one past the last row; 0x2F8 is that plus the column one
// loop starts from).

typedef struct {
  DWORD rva;
  BYTE opcode[2];
  int opcodeLen;
  DWORD offset;
} TableRef;

#define FEATURE_REF_01_RVA 0x6c8986u
static const BYTE FEATURE_REF_01_MARK[5] = { 0xBA, 0xB0, 0x09, 0x09, 0x01 };  // mov edx,table         +0x40's scan
#define FEATURE_REF_02_RVA 0x6c89a4u
static const BYTE FEATURE_REF_02_MARK[6] = { 0x81, 0xFA, 0xA0, 0x0C, 0x09, 0x01 };  // cmp edx,end
#define FEATURE_REF_03_RVA 0x6c89d2u
static const BYTE FEATURE_REF_03_MARK[6] = { 0xFF, 0xB7, 0xBC, 0x09, 0x09, 0x01 };  // push [edi+minLevel]
#define FEATURE_REF_04_RVA 0x6c89d8u
static const BYTE FEATURE_REF_04_MARK[6] = { 0xFF, 0xB7, 0xB8, 0x09, 0x09, 0x01 };  // push [edi+building]
#define FEATURE_REF_05_RVA 0x6c8a23u
static const BYTE FEATURE_REF_05_MARK[5] = { 0xBA, 0xB0, 0x09, 0x09, 0x01 };  // mov edx,table         HasFeature
#define FEATURE_REF_06_RVA 0x6c8a3cu
static const BYTE FEATURE_REF_06_MARK[6] = { 0x81, 0xFA, 0xA0, 0x0C, 0x09, 0x01 };  // cmp edx,end
#define FEATURE_REF_07_RVA 0x6c8a60u
static const BYTE FEATURE_REF_07_MARK[6] = { 0xFF, 0xB6, 0xB8, 0x09, 0x09, 0x01 };  // push [esi+building]
#define FEATURE_REF_08_RVA 0x6c8a69u
static const BYTE FEATURE_REF_08_MARK[6] = { 0x3B, 0x86, 0xBC, 0x09, 0x09, 0x01 };  // cmp eax,[esi+minLevel]
#define FEATURE_REF_09_RVA 0x6c8a92u
static const BYTE FEATURE_REF_09_MARK[5] = { 0xB9, 0xB8, 0x09, 0x09, 0x01 };  // mov ecx,table+8       FeatureOf
#define FEATURE_REF_10_RVA 0x6c8aabu
static const BYTE FEATURE_REF_10_MARK[6] = { 0x81, 0xF9, 0xA8, 0x0C, 0x09, 0x01 };  // cmp ecx,end+8
#define FEATURE_REF_11_RVA 0x6c8ac0u
static const BYTE FEATURE_REF_11_MARK[6] = { 0x8B, 0x80, 0xB0, 0x09, 0x09, 0x01 };  // mov eax,[eax+feature]
#define FEATURE_REF_12_RVA 0x6c8ae2u
static const BYTE FEATURE_REF_12_MARK[5] = { 0xB9, 0xB0, 0x09, 0x09, 0x01 };  // mov ecx,table         BuildingOf
#define FEATURE_REF_13_RVA 0x6c8afbu
static const BYTE FEATURE_REF_13_MARK[6] = { 0x81, 0xF9, 0xA0, 0x0C, 0x09, 0x01 };  // cmp ecx,end
#define FEATURE_REF_14_RVA 0x6c8b12u
static const BYTE FEATURE_REF_14_MARK[6] = { 0x8B, 0x80, 0xB8, 0x09, 0x09, 0x01 };  // mov eax,[eax+building]
#define FEATURE_REF_15_RVA 0x6c8b60u
static const BYTE FEATURE_REF_15_MARK[6] = { 0xFF, 0xB0, 0xB8, 0x09, 0x09, 0x01 };  // push [eax+building]   +0x50
#define FEATURE_REF_16_RVA 0x6c8ba6u
static const BYTE FEATURE_REF_16_MARK[5] = { 0xBA, 0xB0, 0x09, 0x09, 0x01 };  // mov edx,table         +0x54's scan
#define FEATURE_REF_17_RVA 0x6c8bc4u
static const BYTE FEATURE_REF_17_MARK[6] = { 0x81, 0xFA, 0xA0, 0x0C, 0x09, 0x01 };  // cmp edx,end
#define FEATURE_REF_18_RVA 0x6c8bf0u
static const BYTE FEATURE_REF_18_MARK[6] = { 0xFF, 0xB6, 0xB8, 0x09, 0x09, 0x01 };  // push [esi+building]
#define FEATURE_REF_19_RVA 0x6d0728u
static const BYTE FEATURE_REF_19_MARK[5] = { 0xB9, 0xB0, 0x09, 0x09, 0x01 };  // mov ecx,table         RowOf
#define FEATURE_REF_20_RVA 0x6d0744u
static const BYTE FEATURE_REF_20_MARK[6] = { 0x81, 0xF9, 0xA0, 0x0C, 0x09, 0x01 };  // cmp ecx,end

static const TableRef FEATURE_REFS[] = {
  { FEATURE_REF_01_RVA, { 0xBA }, 1, 0x000 }, { FEATURE_REF_02_RVA, { 0x81, 0xFA }, 2, 0x2F0 },
  { FEATURE_REF_03_RVA, { 0xFF, 0xB7 }, 2, 0x00C }, { FEATURE_REF_04_RVA, { 0xFF, 0xB7 }, 2, 0x008 },
  { FEATURE_REF_05_RVA, { 0xBA }, 1, 0x000 }, { FEATURE_REF_06_RVA, { 0x81, 0xFA }, 2, 0x2F0 },
  { FEATURE_REF_07_RVA, { 0xFF, 0xB6 }, 2, 0x008 }, { FEATURE_REF_08_RVA, { 0x3B, 0x86 }, 2, 0x00C },
  { FEATURE_REF_09_RVA, { 0xB9 }, 1, 0x008 }, { FEATURE_REF_10_RVA, { 0x81, 0xF9 }, 2, 0x2F8 },
  { FEATURE_REF_11_RVA, { 0x8B, 0x80 }, 2, 0x000 }, { FEATURE_REF_12_RVA, { 0xB9 }, 1, 0x000 },
  { FEATURE_REF_13_RVA, { 0x81, 0xF9 }, 2, 0x2F0 }, { FEATURE_REF_14_RVA, { 0x8B, 0x80 }, 2, 0x008 },
  { FEATURE_REF_15_RVA, { 0xFF, 0xB0 }, 2, 0x008 }, { FEATURE_REF_16_RVA, { 0xBA }, 1, 0x000 },
  { FEATURE_REF_17_RVA, { 0x81, 0xFA }, 2, 0x2F0 }, { FEATURE_REF_18_RVA, { 0xFF, 0xB6 }, 2, 0x008 },
  { FEATURE_REF_19_RVA, { 0xB9 }, 1, 0x000 }, { FEATURE_REF_20_RVA, { 0x81, 0xF9 }, 2, 0x2F0 },
};
#define FEATURE_REF_COUNT (int)(sizeof FEATURE_REFS / sizeof *FEATURE_REFS)

/** The engine's table as loaded; the first and last rows are the anchor. */
static const FeatureRow *engine_feature_table(void) {
  const FeatureRow *t = (const FeatureRow *)((BYTE *)GetModuleHandleW(NULL) + FEATURE_TABLE_RVA);
  const FeatureRow *first = &t[0], *last = &t[SHIPPED_FEATURE_ROWS - 1];
  if (first->feature != 0 || first->town != 3 || first->building != 17 || first->minLevel != 1) return NULL;
  if (last->feature != 0x2E || last->town != 2 || last->building != 0 || last->minLevel != 4) return NULL;
  return t;
}

static int install_town_features(void) {
  BYTE *base = (BYTE *)GetModuleHandleW(NULL);
  const FeatureRow *engine = engine_feature_table();
  if (!engine) { log_line("town features: the engine's table is not the one we read - nothing done"); return 0; }
  DWORD tableAddress = (DWORD)base + FEATURE_TABLE_RVA;

  // Every reference or none.
  for (int i = 0; i < FEATURE_REF_COUNT; i++) {
    const TableRef *r = &FEATURE_REFS[i];
    const BYTE *at = base + r->rva;
    for (int b = 0; b < r->opcodeLen; b++) {
      if (at[b] != r->opcode[b]) { log_hex("town features: not the bytes we know at ", r->rva); return 0; }
    }
    DWORD have = *(const DWORD *)(at + r->opcodeLen);
    if (have != tableAddress + r->offset) { log_hex("town features: not the table's address at ", r->rva); return 0; }
  }

  int rows = SHIPPED_FEATURE_ROWS + g_ownFeatureCount;
  FeatureRow *ours = (FeatureRow *)VirtualAlloc(NULL, (SIZE_T)rows * sizeof(FeatureRow) + 16, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
  if (!ours) { log_line("town features: no memory for the table"); return 0; }
  for (int i = 0; i < SHIPPED_FEATURE_ROWS; i++) ours[i] = engine[i];
  for (int i = 0; i < g_ownFeatureCount; i++) ours[SHIPPED_FEATURE_ROWS + i] = g_ownFeatures[i];
  DWORD end = (DWORD)(ours + rows);

  for (int i = 0; i < FEATURE_REF_COUNT; i++) {
    const TableRef *r = &FEATURE_REFS[i];
    BYTE *at = base + r->rva + r->opcodeLen;
    // A reference past the last row is the loop's end, and moves with it; a
    // column reference moves with the start.
    DWORD to = r->offset >= 0x2F0 ? end + (r->offset - 0x2F0) : (DWORD)ours + r->offset;
    DWORD old = 0;
    if (!VirtualProtect(at, 4, PAGE_EXECUTE_READWRITE, &old)) { log_hex("town features: could not write at ", r->rva); return 0; }
    *(DWORD *)at = to;
    VirtualProtect(at, 4, old, &old);
    FlushInstructionCache(GetCurrentProcess(), at, 4);
  }
  log_num("town features: the table is ours, rows: ", rows);
  return 1;
}
